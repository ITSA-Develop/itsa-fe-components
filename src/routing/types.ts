import { ComponentType, ReactNode } from 'react';
import { RouteProps } from 'react-router-dom';

import { paths } from './paths';

export type RoutesPathsProp = paths | string;
export type RouteRCTypeProp = ComponentType<any> | null | undefined;

type ProtectedType = { onlyForStaff?: boolean };

type RoutePathsType = { path: RoutesPathsProp | RoutesPathsProp[] };

export type ProtectedRouteProps = RouteProps &
	ProtectedType &
	RoutePathsType & {
		component: ReactNode;
	};

export type PublicRouteProps = RouteProps &
	RoutePathsType & {
		component: ReactNode;
	};
export type BackOfficeRouteProps = RouteProps & RoutePathsType;

export type LayoutComponentProps = {
	component: ReactNode;
};

export type RouteResolverProps = {
	Component: RouteRCTypeProp;
	userData?: any; // TODO: tbd
	isProtected?: boolean;
	isAnonymous?: boolean;
};

export type TLocation = {
	pathname: string;
	hash?: string;
	search?: string;
};

export type ProtectedProps = {
	Component: ReactNode;
	location: TLocation;
} & ProtectedType;
