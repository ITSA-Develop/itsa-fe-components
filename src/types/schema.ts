export type TFieldType = 'text' | 'number' | 'textarea' | 'checkbox';
export type TFieldDisplayType =
	| TFieldType
	| 'dropdown'
	| 'currency'
	| 'radio'
	| 'date'
	| 'dateRange'
	| 'file'
	| 'phone'
	| 'switch';

export interface IFieldSchema {
	type: TFieldDisplayType;
	label?: string;
	rules?: { required: string };
	isDisabled?: boolean;
	hidden?: boolean;
	sorted?: boolean;
	excludeFromFields?: boolean;
	excludeFromColumns?: boolean;
}

export interface IEntitySchema {
	fields: {
		[key: string]: IFieldSchema;
	};
}
