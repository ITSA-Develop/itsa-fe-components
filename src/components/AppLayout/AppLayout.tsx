import { TMenuMode } from '@/types';
import { Layout, LayoutProps, MenuProps } from 'antd';
import { ReactNode, useEffect, useMemo, useCallback } from 'react';
import { HeaderLayout } from './components/HeaderLayout';
import { SidebarLayout } from './components/SidebarLayout';
import { useAppLayoutStore } from '@/store/appLayout.store';
import { dataFromLocalStorage } from '@/helpers/objects';
import { ELocalStorageKeys } from '@/enums';
import { TExtendedMenuItem } from '@/types';
import { useViewportSize } from '@/hooks';

export interface AppLayoutProps extends LayoutProps {
	loading: boolean;
	currentPath: string;
	widthSidebar: number;
	children: ReactNode;
	onClickOptionMenu: (info: { key: string; item: TExtendedMenuItem }) => void;
	userActions?: MenuProps;
	notifications?: MenuProps;
	logo?: string;
	modeSidebar?: TMenuMode;
}

export const AppLayout = ({
	loading,
	currentPath,
	widthSidebar,
	children,
	onClickOptionMenu,
	notifications = { items: [] },
	userActions = { items: [] },
	logo = '',
	modeSidebar = 'inline',
}: AppLayoutProps) => {
	useViewportSize();
	const { setCurrentAgency, setCurrentModule } = useAppLayoutStore();
	const agencies = useAppLayoutStore(state => state.agencies);
	const modulesAgency = useAppLayoutStore(state => state.modulesAgency);

	const selectOrDefault = useCallback(
		<T extends { id: number }>(items: T[], storedId: string | null, setter: (item: T) => void) => {
			if (!items.length) return;

			if (storedId) {
				const parsedId = Number(storedId);
				if (!isNaN(parsedId)) {
					const foundItem = items.find(item => item.id === parsedId);
					if (foundItem) {
						setter(foundItem);
						return;
					}
				}
			}
			setter(items[0]!);
		},
		[],
	);

	const storedIds = useMemo(
		() => ({
			agencyId: dataFromLocalStorage(ELocalStorageKeys.agencyId),
			moduleId: dataFromLocalStorage(ELocalStorageKeys.moduleId),
		}),
		[],
	);

	useEffect(() => {
		selectOrDefault(agencies, storedIds.agencyId, setCurrentAgency);
	}, [agencies, storedIds.agencyId, setCurrentAgency, selectOrDefault]);

	useEffect(() => {
		selectOrDefault(modulesAgency, storedIds.moduleId, setCurrentModule);
	}, [modulesAgency, storedIds.moduleId, setCurrentModule, selectOrDefault]);

	return (
		<div className="flex h-[100dvh] w-full overflow-hidden">
			<Layout className="p-2 gap-2">
				<HeaderLayout loadingHeader={loading} notifications={notifications} userActions={userActions} logo={logo} />
				<SidebarLayout
					width={widthSidebar}
					currentPath={currentPath}
					modeSidebar={modeSidebar}
					onClickOptionMenu={onClickOptionMenu}
					loadingMenu={loading}
				>
					{children}
				</SidebarLayout>
			</Layout>
		</div>
	);
};
