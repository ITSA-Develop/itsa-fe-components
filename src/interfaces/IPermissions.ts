// export interface IGetPermissionResponse {
// 	code: number;
// 	message: string;
// 	result: IPermission;
// }

// export interface IPermission {
// 	agencies: IAgency[];
// }

// export interface IAgency {
// 	id: number;
// 	name: string;
// 	modules: IModule[];
// }

// export interface IModule {
// 	id: number;
// 	name: string;
// 	path: string | null;
// 	icon: string;
// 	entorno: string;
// 	submodules: ISubmodule[];
// }

// export interface ISubmodule {
// 	id: number;
// 	name: string;
// 	path: string | null;
// 	icon: string;
// 	groups?: ISubmodule[];
// 	programs?: ISubmodule[];
// 	actions?: IActions;
// }

// export interface IActions {
// 	all_actions: number;
// 	read: number;
// 	create: number;
// 	update: number;
// 	delete: number;
// }

// // Interfaces para datos transformados
// export interface IAgencyWithStats extends IAgency {
// 	totalModules: number;
// 	modules: (IModule & { totalSubmodules: number })[];
// }

// export interface IPermissionWithStats extends IGetPermissionResponse {
// 	result: {
// 		agencies: IAgencyWithStats[];
// 	};
// }

// export interface IWritableSubmodule extends ISubmodule {
// 	agencyName: string;
// 	moduleName: string;
// }
