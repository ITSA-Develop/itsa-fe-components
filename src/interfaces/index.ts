export interface IGetPermissionResponse {
	code: number;
	message: string;
	result: Result;
}

export interface Result {
	agencies: IAgency[];
}

export interface IAgency {
	id: number;
	name: string;
	modules: IModule[];
}

export interface IModule {
	id: number;
	name: string;
	path: string | null;
	icon: string;
	entorno: string;
	submodules: ISubmodule[];
}

export interface ISubmodule {
	id: number;
	name: string;
	path: string | null;
	icon: string | null;
	url?: string | null;
	groups?: ISubmodule[];
	programs?: ISubmodule[];
	actions?: IActions;
}

export interface IActions {
	all_actions: number;
	read: number;
	create: number;
	update: number;
	delete: number;
}

export interface ISelectOptionDropdownButton {
	items: {
		key: string;
		label: string;
		onClick?: (id: string) => void;
	}[];
}

export interface IValidateRouteResponse {
	code: number;
	message: string;
	result: boolean;
}
//--------------------------------------------------->
// TODO: clean up this file after implementations starts
import { EAddressType, EEmailType, EPhoneConnectionType, EPhoneType } from '@/enums';
import { TDate, TInputOptions, TInputRules } from '@/types';
import { Dayjs } from 'dayjs';
import { ReactNode } from 'react';

export interface IBaseInputProps {
	name: string;
	label: string;
	placeholder?: string;
	rules?: TInputRules;
	error?: string;
	isDisabled?: boolean;
	defaultValue?: string | boolean | number | TInputOptions | TInputOptions[] | null | undefined;
	maxLen?: number;
}

export interface IBaseInputDate extends IBaseInputProps {
	minDate?: Dayjs;
	maxDate?: Dayjs;
	isDisableFuture?: boolean;
	isDisablePast?: boolean;
	shouldOpen?: boolean;
	onClose?: () => void;
	onOpen?: () => void;
}

// TODO: double check with BE what is going to be the Iid returned by default.
export interface IId {
	id: string;
	modified_by: string;
	created_by: string;
	modified_on: TDate;
	created_on: TDate;
}

export interface IManualAddressObject {
	principalStreet: string;
	latitude: number;
	longitude: number;
	postalCode: string;
	province?: number;
	canton?: number;
	parish?: number;
}

export interface ILocationFormData {
	province: string | TInputOptions | null;
	canton: string | TInputOptions | null;
	parish: string | TInputOptions | null;
}

export interface IMultipleEmailItem {
	emailType: EEmailType | string;
	email: string;
	isPrincipal: boolean;
}

export interface IMultiplePhoneItem {
	phoneConnectionType: EPhoneConnectionType | string;
	phoneType: EPhoneType | string;
	phone: string;
	phoneAreaCode: string;
	isPrincipal: boolean;
}

export interface IMultipleAddressItem {
	addressType: EAddressType | string;
	address: {
		principalStreet: string;
		latitude: number;
		longitude: number;
		streetNumber: string;
		provinceId: number;
		cantonId: number;
		parishId: number;
		isManualAddress: boolean;
	};
	isPrincipal: boolean;
	fullAddress: string;
	addressReference: string;
}

export interface IActionPanelOption {
	title: string | ReactNode;
	confirm?: {
		title: string;
		message: string | ReactNode;
	};
	route?: string;
	action?: () => void;
	hidden?: boolean;
	disabled?: boolean;
	target?: string;
}

export interface IPermissionBasic {
	id: number;
	name: string;
}
export interface IPermissionActions {
	update: number;
	delete: number;
	create: number;
	read: number;
	all_actions: number;
}

export interface IPermissionProgram extends IPermissionBasic {
	icon: string;
	actions: IPermissionActions;
	path: string;
	programs?: IPermissionProgram[];
}

export interface IPermissionSubmodule extends IPermissionBasic {
	icon: string;
	path: string;
	programs: IPermissionProgram[];
	groups?: IPermissionProgram[];
}
export interface IPermissionModule extends IPermissionBasic {
	submodules: IPermissionSubmodule[];
}

export interface IPermissionAgency extends IPermissionBasic {
	modules: IPermissionModule[];
}
export interface IPermission extends IPermissionBasic {
	agencies: IPermissionAgency[];
}

export interface IMenuItem {
	title: string;
	icon: string;
	id?: number;
	name?: string;
	path?: string;
	actions?: IPermissionActions;
	groups?: IMenuItem[];
	subList?: IMenuItem[];
}

export interface IUserInfo {
	identification: string;
	identificationType: string;
	name: string;
	picture: string;
	email: string;
	roles: {
		name: string;
	}[];
}

export interface IControllerProps {
	isForm?: boolean;
	isEditable?: boolean;
	isDetail?: boolean;
}

export interface IModalResponsiveProps {
	open: boolean;
	setOpen: (open: boolean) => void;
	onOk: () => void;
	setOnOk: (onOk: () => void) => void;
	onCancel: () => void;
	setOnCancel: (onCancel: () => void) => void;
	title: string;
	setTitle: (title: string) => void;
	footer: ReactNode;
	setFooter: (footer: ReactNode) => void;
	content: ReactNode;
	setContent: (content: ReactNode) => void;
}

export interface ITabsMaintenanceItem {
	key: string;
	label: string;
	children: ReactNode;
	icon?: ReactNode;
}
