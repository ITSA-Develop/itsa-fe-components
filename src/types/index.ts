import { EOrientation, ESize, ESortOrder } from '@/enums';
import { ColumnType } from 'antd/es/table';
import { ReactNode } from 'react';
import { MenuItemType, SubMenuType } from 'antd/es/menu/interface';
import { IActions } from '@/interfaces';
import { FormItemProps } from 'antd/es/form';
import { Control, FieldPath, FieldValues } from 'react-hook-form';
import { InputProps } from 'antd';
export type { Dayjs as TDayjs } from 'dayjs';

//TODO: Probably deprecated starting from here (need to start doing implementations to clean up this)

export type TInputOptions = {
	label: string;
	value: string | number;
};

export type TOrientation = EOrientation.horizontal | EOrientation.vertical; // check key of typeof

export type TModalSize = ESize.large | ESize.medium | ESize.small; // check key of typeof

export type TModalFooterActions = {
	label: string;
	onClick: () => void;
};

export type TModalFooter = {
	linkAction?: { label: string; url: string };
	primaryAction?: TModalFooterActions;
	secondaryAction?: TModalFooterActions;
	isDisabled?: boolean;
	isLoading?: boolean;
};

export type TTableData = { [key: string]: unknown; TABLE__ALERT?: boolean }[];

export type TSortOrder = ESortOrder.asc | ESortOrder.desc; // check key of typeof

export type TInputRules = {
	required?: string;
	validate?: (value: any) => boolean | string;
};

export type TDate = string;

export type TTranslationSchema = {
	schema: string;
	create: string;
	read: string;
	update: string;
	delete: string;
};

export type TButtonTypes = 'button' | 'submit' | 'reset';

export type TError = {
	message: string;
	code: string;
};

export type PermissionAction = keyof IActions; // 'create' | 'read' | 'update' | 'delete' | 'all_actions'

export type TTab = {
	icon: string;
	label: string;
	content: ReactNode;
	hidden?: boolean;
};

export interface TTableColumn {
	key: string;
	label: string;
	render?: (item: any) => React.ReactNode;
	sortable?: boolean;
}

// New types from here
export type TDrawerSize = 'default' | 'large';
export type TTabItem = {
	key: string;
	label: ReactNode;
	children?: ReactNode;
	forceRender?: boolean;
	disabled?: boolean;
	closable?: boolean;
	icon?: ReactNode;
};

export type TNotificationType = 'success' | 'info' | 'warning' | 'error';

export type TMenuMode = 'inline' | 'vertical' | 'horizontal';

// New types from here TABLE
export type TPrimitiveKey<T> = Extract<keyof T, string | number>;

export type TStrictColumnType<T> = Omit<ColumnType<T>, 'dataIndex'> & {
	dataIndex?: TPrimitiveKey<T> | readonly TPrimitiveKey<T>[];
};

export type TStrictTableColumnsType<T> = TStrictColumnType<T>[];

export interface ITableColumnAction<T = any> {
	key: string;
	title: string;
	icon?: ReactNode;
	action: (record: T) => void;
}
export type TMenuItemData = {
	path: string | null;
	icon: React.ReactNode | null;
	url?: string | null;
	actions?: {
		allActions: number;
		read: number;
		create: number;
		update: number;
		delete: number;
	};
	type: 'program' | 'module' | 'submodule' | 'group';
	parentGroup?: string;
	parentSubmodule: string;
	parentModule: string;
};
// Tipo extendido para los items del men√∫ con datos adicionales
export type TExtendedMenuItem = (SubMenuType | MenuItemType) & {
	data?: TMenuItemData;
};

export type Option = { label: React.ReactNode; value: string | number };

export type BaseFieldProps = {
	/** Etiqueta visible (solo UI) */
	label?: React.ReactNode;
	/** Marca visual de requerido (solo UI) */
	requiredMark?: boolean;
	/** Props del Form.Item (layout) */
	formItemProps?: Omit<FormItemProps, 'label' | 'required' | 'validateStatus' | 'help'>;
	/** Ocultar el item */
	hidden?: boolean;
	/** Mensaje de ayuda adicional (UI) */
	extra?: React.ReactNode;
};

export type RHFModeProps<TFieldValues extends FieldValues> = {
	mode: 'rhf';
	/** Nombre del campo (type-safe con FieldPath<T>) */
	name: FieldPath<TFieldValues>;
	/** Control de RHF */
	control: Control<TFieldValues>;
};

export type ControlledModeProps<TValue> = {
	mode: 'controlled';
	value?: TValue;
	onChange?: (v: TValue) => void;
	/** Mensaje de error en modo controlado (opcional) */
	errorMessage?: string;
};

export type RHFInputProps<TFieldValues extends FieldValues> = Omit<InputProps, 'value' | 'onChange'> &
	BaseFieldProps &
	(RHFModeProps<TFieldValues> | ControlledModeProps<string>);

export type TNotificationProps ={
	type: TNotificationType;
	message: string;
	description: string;
}

export type MakeFunctionParamsOptional<F> = F extends (...args: infer P) => infer R ? (...args: Partial<P>) => R : never;






