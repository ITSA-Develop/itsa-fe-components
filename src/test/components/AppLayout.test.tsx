import React from 'react';
import { render, screen, waitFor } from '@testing-library/react';
import { MenuProps } from 'antd';
import { describe, expect, it, vi, beforeEach, afterEach } from 'vitest';
import '@testing-library/jest-dom';
import { AppLayout } from '../../components/AppLayout/AppLayout';
import { TExtendedMenuItem } from '../../types';
import { IAgency, IModule } from '../../interfaces';

// Mock localStorage
const localStorageMock = (() => {
	let store: Record<string, string> = {};
	return {
		getItem: vi.fn((key: string) => store[key] || null),
		setItem: vi.fn((key: string, value: string) => {
			store[key] = value;
		}),
		removeItem: vi.fn((key: string) => {
			delete store[key];
		}),
		clear: vi.fn(() => {
			store = {};
		}),
	};
})();

Object.defineProperty(window, 'localStorage', {
	value: localStorageMock,
});

// Mock the stores
const mockAppLayoutStore = {
	agencies: [] as IAgency[],
	modulesAgency: [] as IModule[],
	setCurrentAgency: vi.fn(),
	setCurrentModule: vi.fn(),
	currentAgency: undefined,
	currentModule: undefined,
	userName: undefined,
	userRole: undefined,
};

vi.mock('../../store/appLayout.store', () => ({
	useAppLayoutStore: () => mockAppLayoutStore,
}));

// Mock helpers
vi.mock('../../helpers/objects', () => ({
	dataFromLocalStorage: vi.fn((key: string) => localStorageMock.getItem(key)),
}));

// Mock component dependencies
vi.mock('../../components/AppLayout/components/HeaderLayout', () => ({
	HeaderLayout: (props: any) => (
		<div data-testid="header-layout" data-props={JSON.stringify(props)}>
			Header Content
		</div>
	),
}));

let mockSidebarOnClick: any = null;

vi.mock('../../components/AppLayout/components/SidebarLayout', () => ({
	SidebarLayout: (props: any) => {
		// Guardamos la función onClickOptionMenu para poder probarla
		mockSidebarOnClick = props.onClickOptionMenu;
		return (
			<div data-testid="sidebar-layout" data-props={JSON.stringify({...props, onClickOptionMenu: 'function'})}>
				{props.children}
			</div>
		);
	},
}));

describe('AppLayout', () => {
	const mockOnClickOptionMenu = vi.fn();
	
	const defaultProps = {
		loading: false,
		currentPath: '/',
		widthSidebar: 280,
		onClickOptionMenu: mockOnClickOptionMenu,
		children: <div>Test Content</div>,
	};

	const mockAgencies: IAgency[] = [
		{
			id: 1,
			name: 'Agencia Principal',
			modules: [
				{
					id: 10,
					name: 'Módulo Ventas',
					path: '/ventas',
					icon: 'shop',
					entorno: 'production',
					submodules: [],
				},
				{
					id: 11,
					name: 'Módulo Inventario',
					path: '/inventario',
					icon: 'inventory',
					entorno: 'production',
					submodules: [],
				},
			],
		},
		{
			id: 2,
			name: 'Agencia Secundaria',
			modules: [
				{
					id: 20,
					name: 'Módulo Compras',
					path: '/compras',
					icon: 'cart',
					entorno: 'production',
					submodules: [],
				},
			],
		},
	];

	const mockModules: IModule[] = [
		{
			id: 10,
			name: 'Módulo Ventas',
			path: '/ventas',
			icon: 'shop',
			entorno: 'production',
			submodules: [],
		},
		{
			id: 11,
			name: 'Módulo Inventario',
			path: '/inventario',
			icon: 'inventory',
			entorno: 'production',
			submodules: [],
		},
	];

	beforeEach(() => {
		vi.clearAllMocks();
		localStorageMock.clear();
		mockAppLayoutStore.agencies = [];
		mockAppLayoutStore.modulesAgency = [];
		mockAppLayoutStore.currentAgency = undefined;
		mockAppLayoutStore.currentModule = undefined;
		mockSidebarOnClick = null;
	});

	afterEach(() => {
		vi.clearAllMocks();
	});

	it('should render correctly with basic props', () => {
		const { container } = render(<AppLayout {...defaultProps} />);

		expect(screen.getByText('Test Content')).toBeInTheDocument();
		expect(screen.getByTestId('header-layout')).toBeInTheDocument();
		expect(screen.getByTestId('sidebar-layout')).toBeInTheDocument();
		expect(container).toMatchSnapshot();
	});

	it('should pass correct props to HeaderLayout', () => {
		const notifications: MenuProps = {
			items: [{ key: 'n1', label: 'Nueva notificación' }],
		};
		const userActions: MenuProps = {
			items: [
				{ key: 'u1', label: 'Perfil' },
				{ key: 'u2', label: 'Cerrar sesión' },
			],
		};
		const logo = 'https://example.com/logo.png';

		render(
			<AppLayout
				{...defaultProps}
				notifications={notifications}
				userActions={userActions}
				logo={logo}
			/>
		);

		const header = screen.getByTestId('header-layout');
		const headerProps = JSON.parse(header.getAttribute('data-props') || '{}');

		expect(headerProps.loadingHeader).toBe(false);
		expect(headerProps.notifications).toEqual(notifications);
		expect(headerProps.userActions).toEqual(userActions);
		expect(headerProps.logo).toBe(logo);
	});

	it('should pass correct props to SidebarLayout', () => {
		const modeSidebar = 'vertical';

		render(
			<AppLayout
				{...defaultProps}
				modeSidebar={modeSidebar}
			/>
		);

		const sidebar = screen.getByTestId('sidebar-layout');
		const sidebarProps = JSON.parse(sidebar.getAttribute('data-props') || '{}');

		expect(sidebarProps.width).toBe(280);
		expect(sidebarProps.currentPath).toBe('/');
		expect(sidebarProps.modeSidebar).toBe(modeSidebar);
		expect(sidebarProps.onClickOptionMenu).toBe('function');
	});

	it('should use default props when optional props are not provided', () => {
		render(<AppLayout {...defaultProps} />);

		const header = screen.getByTestId('header-layout');
		const headerProps = JSON.parse(header.getAttribute('data-props') || '{}');

		expect(headerProps.notifications).toEqual({ items: [] });
		expect(headerProps.userActions).toEqual({ items: [] });
		expect(headerProps.logo).toBe('');

		const sidebar = screen.getByTestId('sidebar-layout');
		const sidebarProps = JSON.parse(sidebar.getAttribute('data-props') || '{}');

		expect(sidebarProps.modeSidebar).toBe('inline');
	});

	it('should call setCurrentAgency when agencies are available and no stored ID', async () => {
		mockAppLayoutStore.agencies = mockAgencies;
		
		render(<AppLayout {...defaultProps} />);

		await waitFor(() => {
			expect(mockAppLayoutStore.setCurrentAgency).toHaveBeenCalledWith(mockAgencies[0]);
		});
	});

	it('should call setCurrentModule when modules are available and no stored ID', async () => {
		mockAppLayoutStore.modulesAgency = mockModules;
		
		render(<AppLayout {...defaultProps} />);

		await waitFor(() => {
			expect(mockAppLayoutStore.setCurrentModule).toHaveBeenCalledWith(mockModules[0]);
		});
	});

	it('should select agency from localStorage when available', async () => {
		mockAppLayoutStore.agencies = mockAgencies;
		localStorageMock.setItem('agencyId', '2');

		render(<AppLayout {...defaultProps} />);

		await waitFor(() => {
			expect(mockAppLayoutStore.setCurrentAgency).toHaveBeenCalledWith(mockAgencies[1]);
		});
	});

	it('should select module from localStorage when available', async () => {
		mockAppLayoutStore.modulesAgency = mockModules;
		localStorageMock.setItem('moduleId', '11');

		render(<AppLayout {...defaultProps} />);

		await waitFor(() => {
			expect(mockAppLayoutStore.setCurrentModule).toHaveBeenCalledWith(mockModules[1]);
		});
	});

	it('should fallback to first item when localStorage ID is invalid', async () => {
		mockAppLayoutStore.agencies = mockAgencies;
		localStorageMock.setItem('agencyId', '999'); // ID que no existe

		render(<AppLayout {...defaultProps} />);

		await waitFor(() => {
			expect(mockAppLayoutStore.setCurrentAgency).toHaveBeenCalledWith(mockAgencies[0]);
		});
	});

	it('should not call setters when arrays are empty', async () => {
		mockAppLayoutStore.agencies = [];
		mockAppLayoutStore.modulesAgency = [];
		
		render(<AppLayout {...defaultProps} />);

		await waitFor(() => {
			expect(mockAppLayoutStore.setCurrentAgency).not.toHaveBeenCalled();
			expect(mockAppLayoutStore.setCurrentModule).not.toHaveBeenCalled();
		});
	});

	it('should handle onClickOptionMenu callback', () => {
		const mockMenuItem: TExtendedMenuItem = {
			key: 'test-key',
			label: 'Test Item',
		};

		render(<AppLayout {...defaultProps} />);

		// Simular el click en una opción del menú usando la función guardada del mock
		const mockInfo = { key: 'test-key', item: mockMenuItem };
		if (mockSidebarOnClick) {
			mockSidebarOnClick(mockInfo);
		}

		expect(mockOnClickOptionMenu).toHaveBeenCalledWith(mockInfo);
	});

	it('should render with loading state', () => {
		render(<AppLayout {...defaultProps} loading={true} />);

		const header = screen.getByTestId('header-layout');
		const headerProps = JSON.parse(header.getAttribute('data-props') || '{}');

		expect(headerProps.loadingHeader).toBe(true);
	});

	it('should handle different currentPath values', () => {
		const testPath = '/dashboard/sales';
		render(<AppLayout {...defaultProps} currentPath={testPath} />);

		const sidebar = screen.getByTestId('sidebar-layout');
		const sidebarProps = JSON.parse(sidebar.getAttribute('data-props') || '{}');

		expect(sidebarProps.currentPath).toBe(testPath);
	});

	it('should handle different sidebar widths', () => {
		const customWidth = 350;
		render(<AppLayout {...defaultProps} widthSidebar={customWidth} />);

		const sidebar = screen.getByTestId('sidebar-layout');
		const sidebarProps = JSON.parse(sidebar.getAttribute('data-props') || '{}');

		expect(sidebarProps.width).toBe(customWidth);
	});

	it('should render layout structure correctly', () => {
		const { container } = render(<AppLayout {...defaultProps} />);

		const layoutWrapper = container.querySelector('.flex.h-\\[100dvh\\].w-full.overflow-hidden');
		expect(layoutWrapper).toBeInTheDocument();

		const layoutContainer = container.querySelector('.p-2.gap-2');
		expect(layoutContainer).toBeInTheDocument();
	});
});
