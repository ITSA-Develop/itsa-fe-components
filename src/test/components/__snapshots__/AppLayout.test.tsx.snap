// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AppLayout > should render correctly with basic props 1`] = `
<div>
  <div
    class="flex h-[100dvh] w-full overflow-hidden"
  >
    <div
      class="ant-layout p-2 gap-2 css-dev-only-do-not-override-l9pxc0"
    >
      <div
        data-props="{"loadingHeader":false,"notifications":{"items":[]},"userActions":{"items":[]},"logo":""}"
        data-testid="header-layout"
      >
        Header Content
      </div>
      <div
        data-props="{"width":280,"currentPath":"/","modeSidebar":"inline","onClickOptionMenu":"function","children":{"type":"div","key":null,"props":{"children":"Test Content"},"_owner":null,"_store":{}}}"
        data-testid="sidebar-layout"
      >
        <div>
          Test Content
        </div>
      </div>
    </div>
  </div>
</div>
`;
